Implement custom dependency injection mechanism:

Mwxyz2022 opened the pull request:
- configured @component and @inject annotations with runtime retention policy
- implemented the core logic for the injector class to create and manage instances
- added a caching mechanism to the injector to store and reuse created instances
- used reflection to instantiate objects based on their constructors
- enabled automatic dependency injection for fields annotated with @inject
- marked service implementations (FileReaderServiceImpl, ProductParserImpl, ProductServiceImpl) with the @Component annotation
- injected dependencies into ProductServiceImpl using the @Inject annotation
- updated the main method to retrieve the ProductService instance via the injector

Mwxyz2022 committed:
Implement custom dependency injection mechanism:

- configured @component and @inject annotations with runtime retention policy
- implemented the core logic for the injector class to create and manage instances
- added a caching mechanism to the injector to store and reuse created instances
- used reflection to instantiate objects based on their constructors
- enabled automatic dependency injection for fields annotated with @inject
- marked service implementations (FileReaderServiceImpl, ProductParserImpl, ProductServiceImpl) with the @Component annotation
- injected dependencies into ProductServiceImpl using the @Inject annotation
- updated the main method to retrieve the ProductService instance via the injector

mateacademy-ai-mentor commented on src/main/java/mate.academy/lib/Injector.java:
```diff
@@ -1,13 +1,68 @@
+        if (!implementationClass.isAnnotationPresent(Component.class)) {
+            throw new RuntimeException("Class " + implementationClass.getName()
+                    + " is not annotated with @Component");
+        }
```
Issue: The code checks if the implementation class is annotated with @Component and throws a RuntimeException if not. According to the checklist, all implementation classes must be annotated with @Component. Please make sure that FileReaderServiceImpl, ProductParserImpl, and ProductServiceImpl are annotated with @Component, otherwise this code will fail at runtime.

mateacademy-ai-mentor reviewed (approved):
### Review Summary

Great job! 🎉 All implementation classes are correctly annotated with `@Component`, and your code fully meets the core requirements and functionality of the task. The Injector checks for the annotation as required, and exception handling is informative and robust. Keep up the good work! If you want to further improve your code, consider reviewing best practices and code style, but there are no critical issues to address at this stage.

***

✨ Please rate this review in the Platform's notification center[^1]. I will be grateful for your feedback! ✨

[^1]: [Rate AI review example](https://snipboard.io/gD6XKF.jpg)

Elena-Bruyako commented on src/main/java/mate.academy/lib/Injector.java:
```diff
@@ -1,13 +1,68 @@
+    private final Map<Class<?>, Object> instances = new HashMap<>();
+    private final Map<Class<?>, Class<?>> interfaceImplementations;
+
+    private Injector() {
```
it's better to use static block of initialization

Elena-Bruyako commented on src/main/java/mate.academy/lib/Injector.java:
```diff
@@ -1,13 +1,68 @@
+            return instances.get(interfaceClazz);
+        }
+
+        Class<?> implementationClass = interfaceImplementations.get(interfaceClazz);
```
create separate method `getImplementation`

Elena-Bruyako commented on src/main/java/mate.academy/lib/Injector.java:
```diff
@@ -1,13 +1,68 @@
+            for (Field field : declaredFields) {
+                if (field.isAnnotationPresent(Inject.class)) {
+                    Object fieldInstance = getInstance(field.getType());
+                    field.setAccessible(true);
```
create method to create new instance through the constructor

Mwxyz2022 committed:
improved injector class structure:

- moved initialization of the interfaceImplementations map to static block
- extracted logic for finding an implementation class into a new private getImplementation method
- created a createNewInstance private method to handle object creation and dependency injection
- simplified the getInstance method by delegating responsibilities to new helper methods

liliia-ponomarenko reviewed (approved):
Good job!
